### ISSUE TITLE
Possibility to opt out of message ordering guarantee in sharding

### ISSUE THREAD

**Issue by johanandren** [akka/akka#30957]
When a shard is being rebalanced between nodes, a shard region tries as best as possible to buffer the messages and then forward to the new final destination, however, there is a scenario where buffered messages needs to be dropped to guarantee that they do not arrive out of order (in alignment with general actor at-most-once and messages arrive in the order they were sent from the same sender).

In some scenarios the ordering guarantee may not be as important, for example if the only place a specific request is sent is from an incoming HTTP request through sharding to the final destination, in such a case a single route may only ever send a single message, and parallel requests from the same HTTP client are anyway not ordered, and the ordering guarantee may therefore be less important than the chance that a message arrives.

It could perhaps be worth some form of opt-out of dropping the buffer in such a scenario, potentially for an explicit subset of messages that will be sent through sharding to the entity.

**Comment by pvlugter** [comment]
Would it also be possible to retain the ordering and forward these messages? By letting the region that's handing off the shard be the first to forward messages, so that any messages it received from other regions are delivered first, before the buffered messages from other regions are sent. It would mean more coordination at the end of handoff, giving the new shard home only to the old region, and some extra delay for the other waiting messages.

**Comment by patriknw** [comment]
It would be difficult to know when those messages have reached the final destination and it's ok to send other buffered messages.

That said, I need to refresh what the dropping condition is here.

**Comment by pvlugter** [comment]
Yes, would need to have some kind of finished message go via the new region to the coordinator/rebalancer. Complicates the handoff a little more.

Currently it looks something like:

- rebalancer sends out the begin-handoff message to all regions
- each region removes the shard, start buffering any new messages, send ack message back
- once all acked, rebalancer sends actual handoff message to the old region
- old region drops any buffered messages (since begin-handoff), in case they came from other regions before they received the begin-handoff, to avoid out-of-order from same sender
- old region sends shard-stopped message to the rebalancer, which completes the rebalance

Instead of dropping any buffered messages, I think there would then need to be more coordination:

- old region notifies the rebalancer that the shard is stopped
- the rebalancer requests the new home for the shard and sends this just to the old region (rebalance still in progress)
- old region forwards any buffered messages to the new region
- to ensure ordering, the old region sends a buffer-flushed message to the new region, which forwards it to the rebalancer
- rebalance can be marked as done now, and other regions can send any buffered messages to the new home

**Comment by johanandren** [comment]
One idea would be to include a limited number of buffered messages in the `ShardStopped` message to the coordinator, and then pass those on in the `HostShard` message from the coordinator to the new region for delivery before responding to the coordinator that the shard stopped, that way we should know that the same sender on another node didn't get its messages reordered.

**Comment by pvlugter** [comment]
Sounds reasonable, and we don't expect many messages to be buffered like this.

**Comment by jrudolph** [comment]
It seems currently the main reason for dropped buffers is that there's a race condition between a shard becoming aware of a shard region being moved and propagating messages in the meantime.

Could we avoid that race condition by sending the acknowledgements on the same path as propagated messages?

 * coordinator send begin-handoff to all regions
   * region containing the shard keeps processing messages to that shard until it gets the handoff message
   * other regions send the ack not back to the coordinator but to the old region (which just forwards to the coordinator)

That way we would never create a buffer on the old region until it starts actually handing off.

Could that work?

**Comment by jrudolph** [comment]
I also like @johanandren's suggestion to use `StopShard` which seems less intrusive.
